### PR TITLE
Vite: Make vite a peer dependency, update plugins

### PR DESCRIPTION
**Pull request by IanVS**
Issue:

Closes https://github.com/storybookjs/storybook/pull/20227
Closes https://github.com/storybookjs/storybook/pull/20214

## What I did

This moves `vite` to a peer dependency, and allows either version 3 or 4.  

It keeps the plugins as normal dependencies.  They are only used as fallbacks in case the user does not have a `vite.config.js`.  For the most part, the newer plugins will work with vite 3, with the possible exception of svelte.  But the solution in those cases is pretty straightforward (update vite or install an older plugin and use a vite.config.js).

As noted in #20227, we may want to add a migration step to add vite, but we don't need to add one for the plugins, if we go with this approach.

## How to test

CI should pass for all vite frameworks.

### CODE REVIEW

**Review by JReinhold — CHANGES_REQUESTED** [review]
Code changes looks good, but we should probably add an entry to MIGRATION.md

We might also want an automigration in this PR? I don't know.

**Review by JReinhold — APPROVED** [review]
Scratch that, vite was already a peer dep in 6.5, no migration nor automigration needed I think

**Comment by IanVS** [comment]
I need to add an eslint ignore because the svelte plugin does not have a main in package.json.

File: code/frameworks/svelte-vite/src/preset.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ export const core: StorybookConfig['core'] = {
 
 export const viteFinal: NonNullable<StorybookConfig['viteFinal']> = async (config, options) => {
   const { plugins = [] } = config;
+  // TODO: set up eslint import to use typescript resolver
+  // eslint-disable-next-line import/no-unresolved
```
Comment (IanVS) [inline code/frameworks/svelte-vite/src/preset.ts:13]: https://github.com/sveltejs/vite-plugin-svelte/issues/552

**Comment by ndelangen** [comment]
https://github.com/storybookjs/storybook/pull/20294 does something similar 👍

File: code/frameworks/react-vite/package.json [hunk 2]
```diff
@@ -55,19 +55,20 @@
...
     "@types/node": "^16.0.0",
-    "typescript": "~4.9.3"
+    "typescript": "~4.9.3",
+    "vite": "^4.0.0"
   },
   "peerDependencies": {
     "react": "^16.8.0 || ^17.0.0 || ^18.0.0",
-    "react-dom": "^16.8.0 || ^17.0.0 || ^18.0.0"
+    "react-dom": "^16.8.0 || ^17.0.0 || ^18.0.0",
+    "vite": "3 - 4"
```
Comment (ndelangen) [inline code/frameworks/react-vite/package.json:71]:
```suggestion
    "vite": "^3.0.0 || ^4.0.0"
```

**Review by ndelangen — APPROVED** [review]

**Comment by IanVS** [comment]
@ndelangen it looks like during one of the merges of `next`, we lost the fix for pinning react-docgen.  I've re-applied it here, and 🤞 that CI will be green now.

**Review by shilman — APPROVED** [review]
LGTM! @IanVS do we need to do a followup with some combination of:
- Documentation updates
- Automigration to automatically add a vite dep if you don't have it

**Comment by IanVS** [comment]
@shilman we could, though I'm not entirely convinced it's necessary.  I think it makes sense to have the migration for react, since it's surprising to need react in a svelte project, but it's not surprising to need vite in a svelte-vite project.  Also, in 6.5, vite was a peer dependency, so this is not a change in 7.0.  My personal recommendation would be to avoid making our already-very-long mirgration doc even longer, and wait to see if we hear any confusion over it.  Personally I don't think we will.

**Comment by shilman** [comment]
@IanVS WFM! Merging 🎉

**Pull request merged**